---
title: "Fragmentomic analysis of cfDNA in sepsis"
author: "Eddie Cano-Gamez"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This code comprises different analyses of fragmentomic characteristics in cfDNA extracted from the plasma of sepsis patients and profiled using TAPS followed by sequencing. The characteristics analysed include fragment size distribution, proportion of cfDNA orginating from different nucleosomal contexts, and analysis of end-motif sequences of cfDNA molecules. 

# Loading libraries
We begin by loading the following R packages
```{r load_libraries, message=FALSE, warning=FALSE}
library(tidyverse)
library(ggpubr)
library(ggrepel)
library(cowplot)
library(reshape2)
library(pheatmap)
library(MatrixGenerics)
```

# Loading data
We next load fragmentomic information obtained after alignment of sequencing data to the human genome. Due to the size of this data and to ensure efficient memory usage, we will load one file at a time, extract the most relevant information, and remove the remaining data from RAM.

We start by defining a set of global variables and parameters, including a list of chromosome names and a vector with all 256 possible DNA 4-mers.
```{r define_global_variables, message=FALSE, warning=FALSE}
# Chromsome names
chromosomes <- paste0("chr",c(1:22,"X","Y","M"))

# DNA 4-mers
bases <- c("A","T","G","C")
kmer_sequences <- expand.grid(bases, bases, bases, bases)
kmer_sequences <- apply(kmer_sequences, MARGIN=1 , paste, collapse="")
```

We next identify and list all files of interest as well as their corresponding sample names
```{r list_input_files, message=FALSE, warning=FALSE}
file_list <- list.files("../../results/fragmentomics/end-motifs", pattern="_GRCh38.tsv$", full.names = T)
sample_list <- gsub("_cfDNA.*$","",gsub("^.*/","",file_list))
```

Finally, we read in one file at a time and estimate key metrics reflecting fragment size distribution and end-motif frequencies
```{r fecth_fragmentomic_metrics, eval=FALSE}
fragment_sizes <- tibble()
end_motif_frequencies <- tibble()
for(i in file_list){
  # Defining sample ID
  sample_id <- gsub("_cfDNA.*$","",gsub("^.*/","",i))
  cat(paste0("\nReading data for ", sample_id, "...\n"))
  
  # Reading data to RAM
  dat <- read_tsv(i)
  
  # Keeping only fragments mapping to chromosomes
  dat <- filter(dat, chr %in% chromosomes)
  
  # Fetching fragment size information
  cat("Fetching fragment size information...\n")
  frag_sizes <- table(dat$length)
  frag_sizes <- tibble(
    "sample_id" = sample_id,
    "length" = as.integer(names(frag_sizes)),
    "fragment_count" = as.integer(frag_sizes)
  )
  frag_sizes$fragment_proportion <- frag_sizes$fragment_count/sum(frag_sizes$fragment_count)
  
  fragment_sizes <- rbind(fragment_sizes, frag_sizes)
  
  # Estimating end motif frequencies
  cat("Fetching end motif information...\n")
  cat("Calculating top strand frequencies...\n")
  motif_freqs_TS <- sapply(kmer_sequences, FUN = function(n){
    sum(dat$end_motif_top_strand == n)
  })
  cat("Calculating bottom strand frequencies...\n")
  motif_freqs_BS <- sapply(kmer_sequences, FUN = function(n){
    sum(dat$end_motif_bottom_strand == n)
  })
  
  motif_freqs <- tibble(
    "sample_id" = sample_id,
    "end_motif" = kmer_sequences,
    "fragment_count_top_strand" = motif_freqs_TS,
    "fragment_count_bottom_stramd" = motif_freqs_BS
  )
  
  end_motif_frequencies <- rbind(end_motif_frequencies, motif_freqs)
  
  cat("Cleaning up...\n")
  rm(dat, frag_sizes, motif_freqs_TS, motif_freqs_BS)
  
  cat("...done!\n")
}
```

Finally, we combine the collected metrics with relevant metadata for each sample, including phenotype and time point of sample collection.
```{r add_metadata_to_fragmetomics_metrics, eval=FALSE}
end_motif_frequencies$phenotype <- "Sepsis"
end_motif_frequencies$phenotype[grepl("^HV",end_motif_frequencies$sample_id)] <- "HV"
end_motif_frequencies$time_point <- as.integer(gsub("^.*_","",end_motif_frequencies$sample_id))
end_motif_frequencies$time_point[grepl("^HV",end_motif_frequencies$sample_id)] <- "HV"

fragment_sizes$phenotype <- "Sepsis"
fragment_sizes$phenotype[grepl("^HV",fragment_sizes$sample_id)] <- "HV"
fragment_sizes$time_point <- as.integer(gsub("^.*_","",fragment_sizes$sample_id))
fragment_sizes$time_point[grepl("^HV",fragment_sizes$sample_id)] <- "HV"
```

```{r loading_fragmentomic_metrics_from_disk, message=FALSE, warning=FALSE, echo=FALSE}
end_motif_frequencies <- read_tsv("../../results/fragmentomics/end-motifs/end-motif-frequencies.tsv")
fragment_sizes <- read_tsv("../../results/fragmentomics/fragmentation-metrics/cfDNA-fragment-sizes.tsv")
```

# Fragmetomic analysis
## Analysis of fragment size distributions
We can now analyse the collated fragmentomic metrics per sample. 

We begin by visualising the fragment size distribution of cfDNA observed in each sample group. The plots below show this information in linear and logarithmic scale, respectively.
```{r plot_fragment_size_distribution_by_phenotype, warning=FALSE, message=FALSE}
ggplot(fragment_sizes, aes(x=length, y=fragment_proportion*100)) +
  geom_line(data=fragment_sizes[fragment_sizes$phenotype=="Sepsis",], aes(group=sample_id), color="steelblue") +
  geom_line(data=fragment_sizes[fragment_sizes$phenotype!="Sepsis",], aes(group=sample_id), color="#737373") +
  ylab("Proportion of fragments (%)") +
  xlab("Fragment length (bp)") +
  xlim(20,600) +
  theme_bw() +
  theme(
    panel.grid.minor  = element_blank()
  )

ggplot(fragment_sizes, aes(x=length, y=log10(fragment_proportion*100))) +
  geom_line(data=fragment_sizes[fragment_sizes$phenotype=="Sepsis",], aes(group=sample_id), color="steelblue") +
  geom_line(data=fragment_sizes[fragment_sizes$phenotype!="Sepsis",], aes(group=sample_id), color="#737373") +
  ylab("log10(Proportion of fragments (%))") +
  xlab("Fragment length (bp)") +
  xlim(20,600) +
  ylim(-4,1) +
  theme_bw() +
  theme(
    panel.grid.minor  = element_blank()
  )
```

The plots below show fragment size distributions stratified by time point (i.e. days since ICU admission), in linear and logarithimic scales.
```{r plot_fragment_size_distribution_by_time_point, warning=FALSE, message=FALSE}
ggplot(fragment_sizes[fragment_sizes$phenotype=="Sepsis",], aes(x=length, y=fragment_proportion*100)) +
  geom_line(aes(group=sample_id, color=as.factor(time_point))) +
  ylab("Proportion of fragments (%)") +
  xlab("Fragment length (bp)") +
  scale_color_manual(values=c("darkblue","steelblue","lightblue")) +
  xlim(20,600) +
  theme_bw() +
  theme(
    panel.grid.minor  = element_blank()
  )

ggplot(fragment_sizes[fragment_sizes$phenotype=="Sepsis",], aes(x=length, y=log10(fragment_proportion*100))) +
  geom_line(aes(group=sample_id, color=as.factor(time_point))) +
  ylab("log10(Proportion of fragments (%))") +
  xlab("Fragment length (bp)") +
  scale_color_manual(values=c("darkblue","steelblue","lightblue")) +
  xlim(20,600) +
  ylim(-4,1) +
  theme_bw() +
  theme(
    panel.grid.minor  = element_blank()
  )
```

### Fragment quantification by nucleosomal context
The plots above show how cfDNA fragment has a periodic size distirbutions, peaking at multiples of 168 bp. These reflect the nucleosomal structure (i.e. a DNA molecule wrapped around a single nucleosomal particle will be ~168 bp long).

Thus, we can bin fragments to distinguish between mono-, di-, and poly-nucleosomal fragments as follows:
```{r plot_fragment_size_distribution_by_nucleosomal_context, warning=FALSE, message=FALSE}
ggplot(fragment_sizes, aes(x=length, y=log10(fragment_proportion*100))) +
  geom_line(data=fragment_sizes[fragment_sizes$phenotype=="Sepsis",], aes(group=sample_id), color="steelblue") +
  geom_line(data=fragment_sizes[fragment_sizes$phenotype!="Sepsis",], aes(group=sample_id), color="#737373") +
  ylab("log10(Proportion of fragments (%))") +
  xlab("Fragment length (bp)") +
  geom_vline(xintercept = 120, linetype="dashed") +
  geom_vline(xintercept = 250, linetype="dashed") +
  geom_vline(xintercept = 420, linetype="dashed") +
  xlim(20,600) +
  ylim(-4,1) +
  theme_bw() +
  theme(
    panel.grid.minor  = element_blank()
  )
```

Based on these definitions, we quantify the number of cfDNA fragments which were part of a mono- a di- or a poly-nuclosomal particle:
```{r estimate_nucleosomal_composition, message=FALSE, warning=FALSE}
fragment_composition <- tibble()
for(i in sample_list){
  dat <- fragment_sizes[fragment_sizes$sample_id==i, ]
  
  free <- sum(dat$fragment_count[dat$length <= 120])/sum(dat$fragment_count)
  mono <- sum(dat$fragment_count[dat$length > 120 & dat$length <= 250])/sum(dat$fragment_count)
  di <- sum(dat$fragment_count[dat$length > 250 & dat$length <= 420])/sum(dat$fragment_count)
  poly <- sum(dat$fragment_count[dat$length > 420])/sum(dat$fragment_count)
  
  fragment_composition <- rbind(
    fragment_composition,
    tibble(
      "sample_id"=i,
      "nucleosome_free"=free,
      "mononucleosomal"=mono,
      "dinucleosomal"=di,
      "polynucleosomal"=poly
    )
    )
}
fragment_composition$phenotype <- "sepsis"
fragment_composition$phenotype[grep("^HV",fragment_composition$sample_id)] <- "HV"

fragment_composition <- melt(fragment_composition)
```

The nucleosomal composition of sepsis and healthy cfDNA samples looks as folllows:
```{r plot_nucleosomal_composition_by_phenotype, message=FALSE, warning=FALSE}
ggplot(fragment_composition, aes(x=phenotype, y=value*100)) +
  geom_boxplot() +
  geom_jitter(width=0.15) +
  stat_compare_means(method="t.test") +
  facet_grid(~variable) +
  xlab("") +
  ylab("Proportion of fragments (%)") +
  theme_bw()
```

### Fragmentation indices
The observations above suggest that sepsis patients have a cfDNA characterised by shorter fragments (mostly mononucleosomal), while healthy volunteers have a higher proportion of higher molecular weight cfDNA. Moreover, the excess of short fragments in sepsis seems to gradually disappear over time and is most prominent at ICU admission.

To corroborate this, we calcualte a fragmentation index (FI) similar to that described in Chan et al., AJHG, 2020. Namely, we divide the number of detected mononucleosomal fragments (168 bp) by the number of dinucleosomal fragments (336 bp). These correspond to the two main peaks in the distribution, as indicated in the following graph:
```{r plot_fragment_size_distribution_fragmentation_index, warning=FALSE, message=FALSE}
ggplot(fragment_sizes, aes(x=length, y=log10(fragment_proportion*100))) +
  geom_line(data=fragment_sizes[fragment_sizes$phenotype=="Sepsis",], aes(group=sample_id), color="steelblue") +
  geom_line(data=fragment_sizes[fragment_sizes$phenotype!="Sepsis",], aes(group=sample_id), color="#737373") +
  ylab("log10(Proportion of fragments (%))") +
  xlab("Fragment length (bp)") +
  geom_vline(xintercept = 168, linetype="dashed") +
  geom_vline(xintercept = 336, linetype="dashed") +
  xlim(20,600) +
  ylim(-4,1) +
  theme_bw() +
  theme(
    panel.grid.minor  = element_blank()
  )
```

```{r estimate_fragmentation_indices, message=FALSE, warning=FALSE}
FIs <- sapply(sample_list, FUN=function(i){
  mononucleosomal_count <- fragment_sizes$fragment_count[fragment_sizes$sample_id==i & fragment_sizes$length == 168]
  dinucleosomal_count <- fragment_sizes$fragment_count[fragment_sizes$sample_id==i & fragment_sizes$length == 336]
  FI <- mononucleosomal_count/dinucleosomal_count
  
  return(FI)
})

fragmentation_indices <- tibble(
  sample_id = sample_list,
  phenotype = c(rep("HV",3),rep("Sepsis",13)),
  time_point = c(rep("HV",3), as.integer(gsub("^.*_","",sample_list[4:16]))),
  fragmentation_index = FIs
)
```


Fragmentation indices are increased in sepsis (reflecting the enrichment of cfDNA in short fragments).
```{r plot_fragmentation_index_per_phenotype, message=FALSE, warning=FALSE}
ggplot(fragmentation_indices, aes(x=phenotype, y=fragmentation_index)) +
  geom_boxplot() +
  geom_jitter(width=0.15, size=3) +
  ylab("Fragmentation index") +
  xlab("") +
  stat_compare_means() +
  theme_bw() +
  theme(
    panel.grid.minor = element_blank()
    )
```

This is mostly driven by samples taken at ICU admission, the earliest time point. In contrast, FIs tend to decrease over time, reflecting a gradual recovery to the typical cfDNA fragment size distribution.
```{r plot_fragmentation_index_per_time_point, message=FALSE, warning=FALSE}
ggplot(fragmentation_indices, aes(x=as.factor(time_point), y=fragmentation_index)) +
  geom_boxplot() +
  geom_jitter(width=0.15, size=3) +
  ylab("Fragmentation index") +
  xlab("Days since ICU admission") +
  stat_compare_means(comparisons = list(c("1","3"),c("1","5"),c("1","HV")), method="t.test") +
  theme_bw() +
  theme(
    panel.grid.minor = element_blank()
  )
```

## Analysis of fragment end-motifs
We used a comination of Samtools and the UCSC toolkit to obtain 5' end motifs for each cfDNA fragment detected during sequencing. Each of these motifs is 4 bp long and contains potential information on the nucleases which generated the fragment in question. In this section we analyse this data to identify differences in the cfDNA end-motif landscape between healthy volunteers and septic patients.

We begin by comparing the 5' end motif sequences derived from the top and bottom reference human genome strands. We expect the sequences derived from both strands to be identical:

```{r plot_end_motif_freqs_top_vs_bottom_strand, message=FALSE, warning=FALSE}
ggplot(end_motif_frequencies, aes(x=fragment_count_top_strand, y=fragment_count_bottom_strand)) +
  geom_point() +
  geom_abline(slope = 1, intercept = 0) +
  xlab("5' end motif frequency (top strand)") +
  ylab("5' end motif frequency (bottom strand)") +
  theme_bw() +
  theme(
    panel.grid.minor = element_blank()
    )
```

Having corroborated that motif frequency estimation agrees between both strands, we combine the counts obtained from both cfDNA fragment ends into one final global count per sequence motif.
```{r combine_top_and_bottom_strand_counts, message=FALSE, warning=FALSE}
end_motif_frequencies <- mutate(
  end_motif_frequencies, 
  end_motif_count = fragment_count_top_strand + fragment_count_bottom_strand
  )
```

Based on these global counts, we calculate the proportion of fragments which show each end-motif sequence in each sample
```{r calculate_end_motif_frequencies_per_samples, message=FALSE, warning=FALSE}
end_motif_frequencies$end_motif_proportion <- unlist(
  lapply(sample_list, FUN=function(i){
    dat <- end_motif_frequencies[end_motif_frequencies$sample_id == i,]
    dat$end_motif_count/sum(dat$end_motif_count)
    })
  )

end_motif_frequencies
```

### Differential end motif abundance
We now directly test for differential end-motif abundance based between healthy volunteers and sepsis cases using a Wilcox Test with FDR-based multiple testing correction for the number of motifs tested.
```{r test_for_differential_motif_abundance, message=FALSE, warning=FALSE}
diff_motif_res <- tibble()
for(i in kmer_sequences){
  dat <- end_motif_frequencies[end_motif_frequencies$end_motif == i,]
  motif_diff <- mean(dat$end_motif_proportion[dat$phenotype == "Sepsis"]) - mean(dat$end_motif_proportion[dat$phenotype == "HV"])
  motif_lfc <- log2(mean(dat$end_motif_proportion[dat$phenotype == "Sepsis"])/mean(dat$end_motif_proportion[dat$phenotype == "HV"]))
  test_res <- wilcox.test(
    x = dat$end_motif_proportion[dat$phenotype == "HV"],
    y = dat$end_motif_proportion[dat$phenotype == "Sepsis"]
  )
  motif_pval <- test_res$p.value
  diff_motif_res <- rbind(
    diff_motif_res,
    tibble("motif"=i, "motif_diff"=motif_diff, "motif_lfc"=motif_lfc,"pval"=motif_pval)
  )
}
diff_motif_res$padj <- p.adjust(diff_motif_res$pval, method = "fdr")
```

As shown by the following volcano plot, a number of motifs appear at higher frequencies either in controls (blue) or in sepsis patients (red):
```{r plot_volcano_motif_frequencies, message=FALSE, warning=FALSE}
ggplot(diff_motif_res, aes(x=motif_lfc, y=-log10(padj))) +
  geom_point() +
  geom_hline(yintercept = -log10(0.05), linetype="dashed") +
  geom_vline(xintercept = 0, linetype="dashed") +
  geom_point(data=diff_motif_res[diff_motif_res$padj < 0.05 & diff_motif_res$motif_lfc > 0,],  color="darkred") +
  geom_point(data=diff_motif_res[diff_motif_res$padj < 0.05 & diff_motif_res$motif_lfc < 0,],  color="steelblue") +
  geom_label_repel(data=diff_motif_res[diff_motif_res$padj < 0.05 & diff_motif_res$motif_lfc > 0.2,], aes(label=motif), color="darkred", size=3) +
  geom_label_repel(data=diff_motif_res[diff_motif_res$padj < 0.05 & diff_motif_res$motif_lfc < -0.2,], aes(label=motif), color="steelblue", size=3) +
  xlab("Log2 fold-change in end-motif proportion") +
  ylab("-log10(P adj)") +
  xlim(-1,1) +
  ylim(0,2) +
  theme_bw()
```

The following heatmap shows proportions for each end motif identified as differentially abundant in this analysis. Motif frequencies are shown here as scaled (i.e. z-scored) values across all samples.
```{r plot_motif_freq_heatmap, message=FALSE, warning=FALSE, fig.width=5, fig.height=7}
motif_freq_matrix <- dcast(end_motif_frequencies[,c(1,4,8)], "end_motif ~ sample_id")
rownames(motif_freq_matrix) <- motif_freq_matrix$end_motif
motif_freq_matrix <- motif_freq_matrix[,-1]

pheatmap(motif_freq_matrix[diff_motif_res$motif[diff_motif_res$padj < 0.05],], show_rownames = T, scale = "row")
```

In general, it would seem that motifs starting with CC seem to be present at a higher frequency in sepsis, while motifs starting in A seem to be decreased in sepsis. 

As an additional visualisation aid to help us interpret this observation, we now combine all motifs identified as significalty more frequent in sepsis and as significantly more frequently in HVs, respectively. We then add up their proportions to estimate the overall contribution of each of these end motif sets to the cfDNA pool.
```{r aggregate_DA_motif_freqs, message=FALSE, warning=FALSE}
top_diff_motifs <- tibble(
  "sample_id" = sample_list,
  "time_point" = as.factor(gsub("^.*_","",gsub("HV.*$","HV",sample_list))),
  "phenotype" = ifelse(grepl("^HV",sample_list),"HV","Sepsis"),
  "HV_preferred_end_motifs" = colSums(motif_freq_matrix[diff_motif_res$motif[diff_motif_res$padj < 0.05 & diff_motif_res$motif_lfc < 0],]),
  "sepsis_preferred_end_motifs" = colSums(motif_freq_matrix[diff_motif_res$motif[diff_motif_res$padj < 0.05 & diff_motif_res$motif_lfc > 0],])
  )
top_diff_motifs <- melt(top_diff_motifs)
```

The combined proportion of these motifs taken together is clearly higher in the corresponding biological group:
```{r plot_aggregated_motif_freq_by_phenotype, message=FALSE, warning=FALSE}
ggplot(top_diff_motifs, aes(x=phenotype, y=value*100)) +
  geom_boxplot() +
  geom_jitter(width=0.15, size = 3) +
  facet_grid(~variable) +
  xlab("") +
  ylab("Combined motif frequency (%)") +
  stat_compare_means() +
  theme_bw()
```

As with the decrease in fragment size, the frequency of sepsis-enriched end motifs also seems to gradually decrease following ICU admission.
```{r plot_aggregated_motif_freq_by_time_point, message=FALSE, warning=FALSE}
ggplot(top_diff_motifs, aes(x=time_point, y=value*100)) +
  geom_boxplot() +
  geom_jitter(width=0.15, size = 3) +
  facet_grid(~variable) +
  xlab("") +
  ylab("Combined motif frequency (%)") +
  stat_compare_means() +
  theme_bw()
```

### Comparison of observed and expected end motif frequencies
In order to assess if any end motifs are more prevalent in cfDNA than would be expected by chance, we compare their measured frequencies with the proportion of each motif in the human genome. 

We first load a matrix of estimated motif frequencies in the genome was estimated for all 256 possible 4-mers, which we derived using BBMap and the GRCh38 version of the reference human genome in FASTA format.
```{r load_motif_frequencies_in_human_genome, message=FALSE, warning=FALSE}
hg38_motif_freqs <- read_tsv("../../data/reference-genome/kmer-counts/GRCh38_4-mer-frequencies_BBMap.tsv", col_names = c("end_motif","count"))
```

We next rank all 4-mer motifs based on their observed frequency in the reference human genome (hg38), in our experiment (global), and in each biological group within our expeirment (i.e. sepsis and HVs).
```{r ranks_end_motifs, message=FALSE, warning=FALSE}
# Human reference genome (hg38)
hg38_motif_ranks <- hg38_motif_freqs[order(hg38_motif_freqs$count, decreasing = T),]
hg38_motif_ranks$rank_hg38 <- 1:nrow(hg38_motif_ranks)

# Our study (full)
global_motif_ranks <- tibble(
  end_motif = rownames(motif_freq_matrix),
  motif_prop = rowMeans(motif_freq_matrix)
  )
global_motif_ranks <- global_motif_ranks[order(global_motif_ranks$motif_prop, decreasing = T),]
global_motif_ranks$rank_global <- 1:nrow(global_motif_ranks)

# Out study (HVs)
HV_motif_ranks <- tibble(
  end_motif = rownames(motif_freq_matrix),
  motif_prop = rowMeans(motif_freq_matrix[,grepl("^HV",colnames(motif_freq_matrix))])
)
HV_motif_ranks <- HV_motif_ranks[order(HV_motif_ranks$motif_prop, decreasing = T),]
HV_motif_ranks$rank_HV <- 1:nrow(HV_motif_ranks)

# Out study (Sepsis)
sepsis_motif_ranks <- tibble(
  end_motif = rownames(motif_freq_matrix),
  motif_prop = rowMeans(motif_freq_matrix[,!grepl("^HV",colnames(motif_freq_matrix))])
)
sepsis_motif_ranks <- sepsis_motif_ranks[order(sepsis_motif_ranks$motif_prop, decreasing = T),]
sepsis_motif_ranks$rank_sepsis <- 1:nrow(sepsis_motif_ranks)

# Final merged set
end_motif_ranks <-  left_join(
  left_join(
    left_join(hg38_motif_ranks[,c(1,3)], global_motif_ranks[,c(1,3)], by = "end_motif"),
    HV_motif_ranks[,c(1,3)], by = "end_motif"
    ),
  sepsis_motif_ranks[,c(1,3)], by = "end_motif"
  )
```

The plot below compares the overall distribution of fragment end motifs in our data (Y) compared to their frequency in the human genome (X). We see an enrichment of motifs starting with CC and a depletion of motigs starting by TT/TA/AT. This agrees well with the observations reported by Chan et al., AJHG; 2020 and suggests that cfDNA is mostly cleaved by DNASE1L3.
```{r plot_motif_ranks_observed_vs_expected, message=FALSE, warning=FALSE}
ggplot(end_motif_ranks, aes(x=-log10(rank_hg38), y=-log10(rank_global))) +
  geom_point() +
  geom_hline(yintercept = -log10(10), linetype="dashed") +
  geom_vline(xintercept = -log10(10), linetype="dashed") +
  geom_label_repel(data=end_motif_ranks[end_motif_ranks$rank_global < 10,], aes(label=end_motif)) +
  geom_label_repel(data=end_motif_ranks[end_motif_ranks$rank_hg38 < 10,], aes(label=end_motif)) +
  theme_bw() +
  theme(
    panel.grid.minor = element_blank()
  )
```

However, motifs thend to rank similarly between HVs and sepsis patients recruited in our study:
```{r plot_motif_ranks_by_phenotype, message=FALSE, warning=FALSE}
ggplot(end_motif_ranks, aes(x=-log10(rank_HV), y=-log10(rank_sepsis))) +
  geom_point() +
  geom_hline(yintercept = -log10(10), linetype="dashed") +
  geom_vline(xintercept = -log10(10), linetype="dashed") +
  geom_label_repel(data=end_motif_ranks[end_motif_ranks$rank_sepsis < 10,], aes(label=end_motif)) +
  theme_bw() +
  theme(
    panel.grid.minor = element_blank()
  )
```


### Sequence diversity analysis
Finally, we look at end motifs from a higher level perspective. To do so, rather than analyse their individual proportions, we assess the overall diversity in cfDNA fragment ends by measuring the Shannon entropy, a widely used measure of diversity and information content.

We first define a function which estimates the Shannon entropy for a set of proportions.
```{r define_getShannonEntropy, message=FALSE, warning=FALSE}
getShannonEntropy <- function(props) {
  -sum(props*log2(props))
}
```

Next, we use this function to estimate Shannon entropies for each sample in our study.
```{r estimate_shannon_entropies, message=FALSE, warning=FALSE}
# Entropy calculation
motif_entropies <- tibble()
for(i in sample_list){
  
  props <- end_motif_frequencies$end_motif_proportion[end_motif_frequencies$sample_id==i]
  entropy <- getShannonEntropy(props)
  
  motif_entropies <- rbind(
    motif_entropies,
    tibble("sample_id"=i, "shannon_entropy"=entropy)
  )
  
}

# Data formatting
motif_entropies <- motif_entropies %>%
  mutate(
    time_point = as.factor(gsub("^.*_","",gsub("HV.*$","HV",sample_id))),
    phenotype = ifelse(grepl("^HV",sample_id),"HV","Sepsis")
  ) %>%
  transmute(
    sample_id,
    phenotype,
    time_point,
    shannon_entropy
  )
```


At first glance, Shannon entropy does not seem to be significantly different between the HVs and sepsis patients.
```{r plot_shannon_entropy_by_phenotype, message=FALSE, warning=FALSE}
ggplot(motif_entropies, aes(x=phenotype, y=shannon_entropy)) +
  geom_boxplot() +
  geom_jitter(width=0.15, size=3) +
  stat_compare_means(method = "t.test") +
  theme_bw() +
  theme(
    panel.grid.minor = element_blank()
    )
```

However, stratifying by time point reveals how Shannon entropy values are significantly higher at the first time point compared to the HV group. Entropy values then decrease over time. 
```{r plot_shannon_entropy_by_time_point, message=FALSE, warning=FALSE}
ggplot(motif_entropies, aes(x=time_point, y=shannon_entropy)) +
  geom_boxplot() +
  geom_jitter(width=0.15, size=3) +
  stat_compare_means(comparisons = list(c("1","3"),c("1","5"),c("1","HV")), method = "t.test") +
  theme_bw() +
  theme(
    panel.grid.minor = element_blank()
  )

```

This can be interpreted as cfDNA fragments in sepsis being slightly more diverse in terms of their end motif sequences, but then gradually returning to their usual bias towards CC- starting fragments.
```{r plot_motif_freq_barplots, message=FALSE, warning=FALSE, fig.width=12, fig.height=5}
ggplot(end_motif_frequencies, aes(x=end_motif, y=log10(end_motif_proportion*100))) +
  geom_line(aes(group=sample_id, color=time_point)) +
  ylab("Proportion of fragments (%)") +
  xlab("End motif") +
  scale_color_manual(values=c("darkblue","steelblue","lightblue","darkred")) +
  theme_bw() +
  theme(
    panel.grid = element_blank(),
    axis.text.x = element_blank(), 
    legend.position = "bottom"
    )

motifs_of_interest <- diff_motif_res$motif[diff_motif_res$padj < 0.05 & diff_motif_res$motif_lfc > 0]
ggplot(end_motif_frequencies, aes(x=end_motif, y=log10(end_motif_proportion*100))) +
  geom_line(aes(group=sample_id, color=time_point)) +
  geom_label_repel(data=end_motif_frequencies[end_motif_frequencies$sample_id == "HV124_3" & end_motif_frequencies$end_motif %in% motifs_of_interest,], aes(label=end_motif)) +
  ylab("Proportion of fragments (%)") +
  xlab("End motif") +
  scale_color_manual(values=c("darkblue","steelblue","lightblue","darkred")) +
  theme_bw() +
  theme(
    panel.grid = element_blank(),
    axis.text.x = element_blank(), 
    legend.position = "bottom"
    )

motifs_of_interest <- diff_motif_res$motif[diff_motif_res$padj < 0.05 & diff_motif_res$motif_lfc < 0]
ggplot(end_motif_frequencies, aes(x=end_motif, y=log10(end_motif_proportion*100))) +
  geom_line(aes(group=sample_id, color=time_point)) +
  geom_label_repel(data=end_motif_frequencies[end_motif_frequencies$sample_id == "HV124_3" & end_motif_frequencies$end_motif %in% motifs_of_interest,], aes(label=end_motif)) +
  ylab("Proportion of fragments (%)") +
  xlab("End motif") +
  scale_color_manual(values=c("darkblue","steelblue","lightblue","darkred")) +
  theme_bw() +
  theme(
    panel.grid = element_blank(),
    axis.text.x = element_blank(), 
    legend.position = "bottom"
    )
```

# Exporting results
Finally, we write the results from this analysis into TSV files.
```{r export_results, eval=FALSE}
write.table(fragment_sizes, file = "./results/fragmentomics/cfDNA-fragment-sizes.tsv", sep="\t", quote=F, row.names=F)
write.table(fragmentation_indices, file = "./results/fragmentomics/fragmentation-indices.tsv", sep="\t", quote=F, row.names=F)
write.table(end_motif_frequencies, file = "./results/fragmentomics/end-motif-frequencies.tsv", sep="\t", quote=F, row.names=F)
```

# R session
```{r r_session, message=FALSE, warning=FALSE}
sessionInfo()
```
